packages/ERTP/src/amountMath.js

```diff
@@ -0,0 +1,144 @@
+// so, it must know how to add and subtract digital assets. Rather
+// than hard-coding a particular solution, we chose to parameterize
+// the issuer with a collection of polymorphic functions, which we
+// call `amountMath`. These math functions include concepts like
+// addition, subtraction, and greater than or equal to.
+
+// We also want to make sure there is no confusion as to what kind of
+// asset we are using. Thus, amountMath includes checks of the
+// `brand`, the unique identifier for the type of digital asset. If
+// the wrong brand is used in amountMath, an error is thrown and the
+// operation does not succeed.
+
+// amountMath uses mathHelpers to do most of the work, but then adds
+// the brand to the result. The function `extent` gets the extent from
+// the amount by removing the brand (amount -> extent), and the function
+// `make` adds the brand to produce an amount (extent -> amount). The
+// function `coerce` takes an amount and checks it, returning an amount (amount
+// -> amount).
+
+// `makeAmount` takes in a brand and the name of the particular
+// mathHelpers to use.
+
+// amountMath is unfortunately not pass-by-copy. If you call
+// `getAmountMath` on a remote issuer, it will be a remote object and
+// each call will incur the costs of calling a remote object. However,
+// you can create a local amountMath by importing this module locally
+// and recreating by passing in a brand and an mathHelpers name, both
+// of which can be passed-by-copy (since there are no calls to brand
+// in this module).
+
+// Each issuer of digital assets has an associated brand in a one-to-one
+// mapping. In untrusted contexts, such in analyzing payments and
+// amounts, we can get the brand and find the issuer which matches the
+// brand. The issuer and the brand mutually validate each other.
+
+function makeAmountMath(brand, mathHelpersName) {
+  mustBeComparable(brand);
+  mathHelpersName = `${mathHelpersName}`;
+
+  const helpers = mathHelpersLib[mathHelpersName];
+  assert(helpers !== undefined, 'helpers must be defined');
+
+  // Cache the amount if we can.
+  const cache = new WeakSet();
+
+  const amountMath = harden({
+    getBrand: () => brand,
+    getMathHelpersName: () => mathHelpersName,
+
+    // Make an amount from an extent by adding the brand.
+    make: allegedExtent => {
+      helpers.doAssertKind(allegedExtent);
+      const amount = harden({
+        brand,
+        extent: allegedExtent,
+      });
+      cache.add(amount);
+      return amount;
+    },
+
+    // Make sure this amount (or extent) are valid and return them if so.
+    coerce: allegedAmountOrExtent => {
+      // If the cache already has the allegedAmountOrExtent, that
+      // means it is a valid amount.
+      if (cache.has(allegedAmountOrExtent)) {
+        return allegedAmountOrExtent;
+      }
+      const { brand: allegedBrand, extent } = allegedAmountOrExtent;
+      // If brand is undefined, this isn't an amount, but it could be
```

Comment:
Previously we did this check differently. Not sure if it's better or worse, since it assumes that anything that has a property called 'extent' is trying to be units.

```
      if (!Object.prototype.hasOwnProperty.call(allegedUnits, 'extent')) {
        // This is not a units. Let's see if it's a extent. Will
        // throw on inappropriate extent.
        return unitOps.make(allegedUnits);
      }
```

Reply:
Both commit the same sin. I don't know which is more likely to cause accidental collision, so I'm indifferent between them. Resolving.